src/Entities/Player.cpp

```diff
@@ -339,6 +314,79 @@ void cPlayer::Tick(std::chrono::milliseconds a_Dt, cChunk & a_Chunk)
 
 
 
+void cPlayer::TickFreezeCode(bool a_MyChunkIsSent)
+{
+	// This function is ticked by the player's client handle. This ensures it always ticks, even if the player
+	// is standing in an unloaded chunk, unlike cPlayer::Tick. We need this because the freeze handling code must
+	// also tick in unloaded chunks.
+	if (m_IsFrozen)
+	{
+		m_FreezeCounter += 1;
+		if ((!m_IsManuallyFrozen) && (a_MyChunkIsSent))
+		{
+			cWorld::cLock Lock(*GetWorld());
+			// If the player was automatically frozen, unfreeze if the chunk the player is inside is loaded
+			Unfreeze();
+
+			// Pull the player out of any solids that might have loaded on them.
+			PREPARE_REL_AND_CHUNK(GetPosition(), *(GetParentChunk()));
+			if (RelSuccess)
+			{
+				int NewY = Rel.y;
+				if (NewY < 0)
+				{
+					NewY = 0;
+				}
+				while (NewY < cChunkDef::Height - 2)
+				{
+					// If we find a position with enough space for the player
+					if (
+						(Chunk->GetBlock(Rel.x, NewY, Rel.z) == E_BLOCK_AIR) &&
+						(Chunk->GetBlock(Rel.x, NewY + 1, Rel.z) == E_BLOCK_AIR)
+					)
+					{
+						// If the found position is not the same as the original
+						if (NewY != Rel.y)
+						{
+							SetPosition(GetPosition().x, NewY, GetPosition().z);
+							GetClientHandle()->SendPlayerPosition();
+						}
+						break;
+					}
+					++NewY;
+				}
+			}
+		}
+		else
+		{
+			// If the player was externally / manually frozen (plugin, etc.) or if the chunk isn't loaded yet:
+			// 1. Set the location to m_FrozenPosition every tick.
+			// 2. Zero out the speed every tick.
+			// 3. Send location updates every 60 ticks.
+
+			if ((m_FreezeCounter % 60 == 0) || ((m_FrozenPosition - GetPosition()).SqrLength() > 2 * 2))
```

Comment:
> I continously send the position to the player to prevent slow sinking and to prevent player movement in loaded chunks when a plugin asks for that. I do not employ any client-side freezing.

Forgot about plugins, whoops. Nevertheless, you can still drop both FrozenPosition and FreezeCounter. Use the world age to periodically resend position, and check in cPlayer::SetPosition/SetSpeed that the player is not frozen before setting.

I'm not sure about client freezing, but if someone knows how we can get rid of periodic resending.

nah, slow sink is fine; we both agree on sending the correct position (I propose topmost block) upon unfreeze due to chunk loaded, so no stuck in caves. And if there is a Y height bug, well, let's fix it :D